src/transforms/remap.rs

```diff
@@ -48,16 +53,37 @@ impl TransformConfig for RemapConfig {
     }
 }
 
-#[derive(Debug, Clone)]
+#[derive(Clone)]
 pub struct Remap {
     program: Program,
     timezone: TimeZone,
     drop_on_error: bool,
     drop_on_abort: bool,
+    enrichment_tables: Arc<ArcSwap<HashMap<String, Box<dyn EnrichmentTable + Send + Sync>>>>,
+}
+
+lazy_static::lazy_static! {
+    static ref MUTEX: std::sync::Mutex<usize> = std::sync::Mutex::new(0);
 }
 
 impl Remap {
-    pub fn new(config: RemapConfig) -> crate::Result<Self> {
+    pub fn new(
+        config: RemapConfig,
+        enrichment_tables: Arc<ArcSwap<HashMap<String, Box<dyn EnrichmentTable + Send + Sync>>>>,
+    ) -> crate::Result<Self> {
+        // Ensure we don't have multiple threads running this code at the same time.
+        let lock = MUTEX.lock().unwrap();
+
+        // Add a dummy index to test it works.
+        let mut tables = enrichment_tables.swap(Default::default());
+        match Arc::get_mut(&mut tables).unwrap().get_mut("file") {
+            None => (),
+            Some(table) => table.add_index(vec!["field1"]),
+        }
+        enrichment_tables.swap(tables);
+
+        drop(lock);
```

Comment:
An issue with ArcSwap, as far as I can tell, there is no way to get a mutable reference to the data inside the ArcSwap. It's an all or nothing swap of the contents.

When building up the EnrichmentTables the Remap transform needs to be able to modify the tables so they can build indexes into the data. This is the only way I can see to do it, swap the data out, modify it and then swap it back in again.

It seems to work, and it is happening at load time here, so performance is not a major issue, but it feels dirty!

Reply:
If it's not hot-path and we then primarily just _read_ these tables I'm content with this.